lib/common/strings.c

```diff
@@ -782,3 +783,66 @@ pcmk_numeric_strcasecmp(const char *s1, const char *s2)
     }
     return 0;
 }
+
+int
+pcmk__strcmp(const char *s1, const char *s2, uint32_t flags)
+{
+    /* If this flag is set, the second string is a regex. */
+    if (is_set(flags, pcmk__str_regex)) {
+        regex_t *r_patt = calloc(1, sizeof(regex_t));
+        int reg_flags = REG_EXTENDED | (is_set(flags, pcmk__str_casei) ? REG_ICASE : 0);
+        int regcomp_rc = 0;
+        int rc = 0;
+        int nregs = 0;
+        regmatch_t *pmatch = NULL;
+
+        regcomp_rc = regcomp(r_patt, s2, reg_flags);
+        if (regcomp_rc != 0) {
+            rc = 1;
+            crm_err("Bad regex '%s' for update: %s", s2, strerror(regcomp_rc));
+        } else {
+            rc = regexec(r_patt, s1, nregs, pmatch, 0);
+        }
+
+        free(pmatch);
+        regfree(r_patt);
+        free(r_patt);
+
+        /* regexec returns 0 for a match, and some constant for no match.  We can just
+         * repurpose that non-zero return code here.  It doesn't line up exactly with
+         * how strcmp works, but the caller will know they used a regex and can expect
+         * this return code.
+         */
+        return rc;
+    }
+
+    /* If the strings are the same pointer, return 0 immediately. */
+    if (s1 == s2) {
+        return 0;
+    }
+
+    /* If this flag is set, return 0 if either (or both) of the input strings
+     * are NULL.  If neither one is NULL, we need to continue and compare
+     * them normally.
+     */
+    if (is_set(flags, pcmk__str_null_matches)) {
+        if (s1 == NULL || s2 == NULL) {
+            return 0;
+        }
+    }
+
+    /* Handle the cases where one is NULL and the str_null_matches flag is not set.
+     * A NULL string always sorts to the beginning.
+     */
+    if (s1 == NULL && s2 != NULL) {
+        return -1;
+    } else if (s1 != NULL && s2 == NULL) {
+        return 1;
+    }
```

Comment:
You don't actually need the != NULL parts since we know s1 != s2 at this point

Reply:
Fixed.